### PR TITLE
Add basic int ops to tinybang

### PR DESCRIPTION
**Pull request by maz**
This PR adds some integer less than comparison, and integer subtraction and multiplication to tinybang. This is also the first Ocaml code that I've ever written, so please tell me about any stylistic improvements that I can make.

**Reviewers:** @leafac @rpglover64 @zepalmer

### CODE REVIEW

**Comment by zepalmer** [comment]
Your coding style looks good and I don't see anything wrong with what you've written.  I do, however, notice that some code has yet to be written.  This gap can be identified by failure testing, so some more unit tests are probably in order.  Consider the following toploop session:

```
TinyBang 0.3 Toploop
--------------------

Please enter an expression to evaluate followed by ";;".

x = 1;
y = ();
z = int- y x;;

Fatal error: exception Tiny_bang_interpreter.Evaluation_failure("int- defined only on ints.")
```

Evaluation failures should never come out of the toploop unless the type system has been disabled with `-T`.  Your implementation of the evaluation of `int-` is correct, but the type system has no closure rules for it.  As a result, the type system does not observe the type error on the third line of code above.

The appropriate file to contain this logic is `tiny_bang_constraint_closure.ml` in the type system library.  The current implementations of closure for `int+` and `int=` are written as two separate functions.  That said, it might be a good idea to have a master function which cases on the built-in operator in question; that would, in the future, generate appropriate warnings if constraint closure code had not been written for a built-in operator.

Care to take a swing at that?

**Comment by maz** [comment]
@zepalmer Updated the commit.

File: src/tiny-bang-typechecker/tiny_bang_constraint_closure.ml [hunk 1]
```diff
@@ -228,6 +228,17 @@ let close_by_int_equality (cs : Constraint_database.t) : tbconstraint Enum.t =
...
+(* This exists to not be a total function if someone forgets to write a closure
+   rule.*)
+let builtin_closure op =
+  match op with
+  | Op_int_plus -> close_by_int_op Op_int_plus
+  | Op_int_minus -> close_by_int_op Op_int_minus
+  | Op_int_times -> close_by_int_op Op_int_times
+  | Op_int_equal -> close_by_int_bool_op Op_int_equal
+  | Op_int_lessthan -> close_by_int_bool_op Op_int_lessthan
+  | Op_ref -> raise (Failure "There aren't closure rules for Op_ref")
```
Comment (maz) [inline src/tiny-bang-typechecker/tiny_bang_constraint_closure.ml:240]: I couldn't find any closure rules for `Op_ref`. Is that intended?

Comment (zepalmer) [inline src/tiny-bang-typechecker/tiny_bang_constraint_closure.ml:240]: Nope; that's a bug.  I'll put in an issue for that.

**Comment by zepalmer** [comment]
Looks good to me.  The handling of `Op_ref` needs to be fixed, of course, but that's a separate problem.  In terms of the integer operations, it looks like we're good.

**Comment by zepalmer** [comment]
Clearly, I have no idea how to use this website.  :)

File: src/tiny-bang-ast/tiny_bang_ast.ml [hunk 2]
```diff
@@ -149,14 +152,14 @@ type expr = Expr of ast_uid * clause list
 and clause = Clause of ast_uid * var * redex
 
 (** The type of a reducable expression. *)
-and	redex =
```
Comment (leafac) [inline src/tiny-bang-ast/tiny_bang_ast.ml:152]: I see you like sneaking in some code gardening on your commits. I appreciate the style :+1:

File: src/tiny-bang-interpreter/tiny_bang_interpreter.ml [hunk 3]
```diff
@@ -239,18 +237,33 @@ let builtin_op env op list_var =
...
+
+(*metatheoretic functions for the builtin operators *)
+let builtin_op env op list_var =
+  match op with
+  | Op_int_plus ->
+    builtin_int_binary_op env op list_var (+)
+  | Op_int_minus ->
+    builtin_int_binary_op env op list_var (-)
+  | Op_int_times ->
+    builtin_int_binary_op env op list_var ( * ) (* The spaces ensure that it's not parsed as a comment *)
```
Comment (leafac) [inline src/tiny-bang-interpreter/tiny_bang_interpreter.ml:262]: Why does OCaml do that? :hankey:

**Comment by leafac** [comment]
@maz Great PR. Congratulations on your first contribution to Big Bang.

---

On an unrelated note: you're listed as a collaborator on https://github.com/JHU-PL-Lab/big-bang/, still you sent this PR from a fork https://github.com/maz/big-bang.

Is there any reason why you did that, instead of using a branch in the https://github.com/JHU-PL-Lab/big-bang/ directly? This is not a problem, I'm just curious.

**Comment by maz** [comment]
The workflow that I'm used to is forking and making PRs from that fork, so
that's why I opted to use that.

**Pull request merged**